FEATURE: Implement 2factor login TOTP

featheredtoast opened the pull request:
https://meta.discourse.org/t/two-factor-local-login-proposal/76689/1

TOTP support only. Prompts for a second factor ("2-step authentication") in the following areas:

- Local Login
- Admin Login
- Password Reset
- Email Change

SamSaffron commented on app/controllers/users_controller.rb:
```diff
@@ -146,6 +146,29 @@ def username
     end
   end
 
+  def toggle_second_factor
```
I am not a fan of "toggle" actions that rely on state, I also feel user controller is getting quite big, perhaps a second_factor controller with an update method and then caller tells it if it wants it enable or disabled.

SamSaffron commented on app/controllers/users_controller.rb:
```diff
@@ -146,6 +146,29 @@ def username
+      user.user_second_factor.enabled = true
+      user.user_second_factor.save!
+    end
+    render json: {}
```
this either worked or not ... it should probably send through json_success like other spots do

SamSaffron commented on app/controllers/session_controller.rb:
```diff
@@ -233,6 +238,14 @@ def create
 
+    if user.totp_enabled?
+      unless user.authenticate_otp(params[:second_factor_token])
+        secure_session["second-factor-#{request.remote_ip}"] = "true"
```
this feels risky cause ip can change between requests especially in corporates behind a big NAT

featheredtoast replied:
The only part where this is used is to identify which rate limiter to use -- the "login" one or the "second factor token" one.

An alternative to using this method is to detect whether the second_factor_token parameter is in the request. It could trivially give attackers a few more attempts against a different rate limiter, but it would still not allow a crazy amout of brute forcing.

SamSaffron commented on app/assets/javascripts/discourse/models/user.js.es6:
```diff
@@ -304,6 +304,23 @@ const User = RestModel.extend({
   },
 
+  loadSecondFactorCodes(password) {
+    return ajax("/session/second-factor", {
```
maybe this also goes into the second factor controller

SamSaffron commented:
hmm the name user_second_factor is a little bit strange to me, not sure ... maybe we should go with `user_totp` if this is intended just for totp? when we implement say fido is it going to go into user_second_factors?

SamSaffron commented:
also can you make sure that there is a test covering every new controller action you added?

eviltrout commented on app/assets/javascripts/admin/controllers/admin-user-index.js.es6:
```diff
@@ -41,6 +41,16 @@ export default Ember.Controller.extend(CanCheckEmails, {
   },
 
+  @computed('model.second_factor_enabled')
+  show_second_factor(second_factor_enabled) {
```
In Javascript we use camelCase style for object properties. This should be `showSecondFactor` and `displaySecondFactorEnabled`.

Having said that, `showSecondFactor` is basically an alias to a model property.  It wouldn't be bad to just have `{{#if model.second_factor_enabled}}` in your template.

eviltrout commented on app/assets/javascripts/admin/controllers/admin-user-index.js.es6:
```diff
@@ -41,6 +41,16 @@ export default Ember.Controller.extend(CanCheckEmails, {
+  },
+
+  @computed('model.second_factor_enabled')
+  display_second_factor_enabled(second_factor_enabled) {
```
I feel like this might be better put in the template, that way you don't have to disable escaping.

```hbs
{{#if model.second_factor_enabled}}
  {{i18n "yes_value"}}
{{else}}
  {{i18n "no_value"}}
{{/if}}
```

eviltrout commented on app/assets/javascripts/admin/models/admin-user.js.es6:
```diff
@@ -168,6 +168,17 @@ const AdminUser = Discourse.User.extend({
   },
 
+  disableSecondFactor() {
+    const self = this;
```
We are migrating away from `self = this` as a pattern. Instead use ES6 arrow functions so that `this` points at the correct thing.

eviltrout commented on app/assets/javascripts/admin/models/admin-user.js.es6:
```diff
@@ -168,6 +168,17 @@ const AdminUser = Discourse.User.extend({
+    return ajax("/admin/users/" + this.get('id') + "/disable_second_factor", {
+      type: 'PUT'
+    }).then(function() {
+      self.setProperties({
```
If you're only updating one property it's preferable to use `this.set('second_factor_enabled', true)` rather than `setProperties`

eviltrout commented on app/assets/javascripts/discourse/controllers/login.js.es6:
```diff
@@ -23,14 +24,23 @@ export default Ember.Controller.extend(ModalFunctionality, {
   canLoginLocal: setting('enable_local_logins'),
   loginRequired: Em.computed.alias('application.loginRequired'),
 
+  @observes('secondFactorRequired')
```
We try to avoid observers wherever possible.

In this case, you know exactly when you are updating `secondFactorRequired` - you could just call this function when you do! Or even have a function `updateSecondFactorRequired(required)` that would set it too.

eviltrout commented on app/assets/javascripts/discourse/controllers/password-reset.js.es6:
```diff
@@ -45,7 +47,17 @@ export default Ember.Controller.extend(PasswordValidation, {
+          if (result.errors && result.errors.second_factor) {
+            this.set('secondFactorRequired',true);
+            this.set('password', null);
+            this.set('errorMessage', 'Invalid token');
```
This needs to be translated. You should also use `setProperties` here as you are setting 3 in a row.

eviltrout commented on app/assets/javascripts/discourse/controllers/password-reset.js.es6:
```diff
@@ -45,7 +47,17 @@ export default Ember.Controller.extend(PasswordValidation, {
+            this.set('password', null);
+            this.set('errorMessage', 'Invalid token');
+          }
+          else if (this.get('secondFactorRequired') === true) {
```
Minor but `this.get('secondFactorRequired')` should be sufficient here without `=== true`

eviltrout commented on app/assets/javascripts/discourse/controllers/preferences/second-factor.js.es6:
```diff
@@ -0,0 +1,66 @@
+
+  loading: false,
+  password: null,
+  second_factor_image: null,
```
These properties should be camelCase too.

eviltrout commented on app/assets/javascripts/discourse/controllers/preferences/second-factor.js.es6:
```diff
@@ -0,0 +1,66 @@
+
+  @computed('loading')
+  submitButtonText(loading) {
+    if (loading) return "Loading";
```
This text needs to be translated via I18n

eviltrout commented on app/assets/javascripts/discourse/controllers/preferences/second-factor.js.es6:
```diff
@@ -0,0 +1,66 @@
+        this.set('second_factor_key', resp.key);
+        this.set('second_factor_image', resp.qr);
+      })
+        .catch(popupAjaxError)
```
This indentation is odd here. I'd prefer if `.catch` was on the previous line following the paren.

eviltrout commented on app/assets/javascripts/discourse/models/user.js.es6:
```diff
@@ -304,6 +304,23 @@ const User = RestModel.extend({
+  toggleSecondFactor(token) {
+    return ajax(userPath(`${this.get('username_lower')}/preferences/second-factor`), {
+      dataType: 'json',
+      data: { token: token },
```
Minor but in ES6 this can just be ` { token } `

eviltrout commented on app/assets/javascripts/discourse/routes/preferences-second-factor.js.es6:
```diff
@@ -0,0 +1,21 @@
+import RestrictedUserRoute from "discourse/routes/restricted-user";
+
+export default RestrictedUserRoute.extend({
+  model: function() {
```
instead of `model: function()` you should use the new syntax of `model()`, ditto for the other functions in this class.

eviltrout commented on app/assets/javascripts/discourse/templates/modal/login.hbs:
```diff
@@ -32,6 +32,21 @@
             </tr>
           </table>
         </div>
+        <div id="second-factor" style="display: none;">
```
This part of the template seems to be duplicated. You should use an ember component to avoid repeating yourself.

eviltrout commented on app/assets/javascripts/discourse/templates/password-reset.hbs:
```diff
@@ -16,20 +16,28 @@
+          <div class="input">
+            {{input value=secondFactor id="second-factor" autofocus="autofocus"}}
+          </div>
+          <button class='btn btn-primary' {{action "submit"}}>Submit</button>
```
This text is not translated. Also this should probably be a `{{d-button}}`

eviltrout commented on app/controllers/session_controller.rb:
```diff
@@ -271,6 +284,22 @@ def current
+    RateLimiter.new(nil, "login-min-#{request.remote_ip}", SiteSetting.max_logins_per_ip_per_minute, 1.minute).performed!
+    if user = User.find_by_username_or_email(params[:login])
+      unless user.confirm_password?(params[:password])
+        invalid_credentials
```
Minor but why not `return invalid_credentials`? One line shorter and reads better

eviltrout commented on app/controllers/session_controller.rb:
```diff
@@ -271,6 +284,22 @@ def current
+        return
+      end
+      qrcode = RQRCode::QRCode.new(user.provisioning_uri)
+      qrcode_svg = qrcode.as_svg(offset: 0, color: '000',
```
Formatting is a bit weird here. This line has two keys and the others have one. One style I enjoy is like this:

```ruby
some_method(
  offset: 0,
  color: '000'
)
```

That way if the method name changes or you assign it to a variable you don't need to change the indentation of all following lines which pollutes the git history.

eviltrout commented on app/models/user.rb:
```diff
@@ -1002,6 +1007,33 @@ def from_staged?
     custom_fields[User::FROM_STAGED]
   end
 
+  def totp
```
Our `User` class is already quite big. At the very least I think we might want to put this in a mixin and include it.

We could also consider creating a helper class like `UserTotp.new(user)` then call `create`, `provision_url`, etc on it.

eviltrout commented on app/models/user_second_factor.rb:
```diff
@@ -0,0 +1,3 @@
+class UserSecondFactor < ActiveRecord::Base
+  belongs_to :user, class_name: 'User'
```
`class_name` should not be necesary here.

featheredtoast commented:
Thank you both for the reviews! I'll start working on the recommendations for reactors and add in the missing tests. Wanted to get the working prototype and architecture approved before building out test expectations too much.

SamSaffron commented:
Cool @featheredtoast ... where are we now, how close to having this mergable?

featheredtoast commented:
Hi Sam, sorry if it wasn't clear from last week's update - I consider this
mergable, pending any other feedback the team might have.

I rebased last week, but can do one more to get things up to date if you
are ready for it now.

On Feb 12, 2018 6:56 PM, "Sam" <notifications@github.com> wrote:

> Cool @featheredtoast <https://github.com/featheredtoast> ... where are we
> now, how close to having this mergable?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/discourse/discourse/pull/5521#issuecomment-365135499>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQuJoLi0gHH-vMw2fDT0z9dtCsB9Iirks5tUPnGgaJpZM4Ro7il>
> .
>

featheredtoast commented:
Just rebased/squashed and added 2-factor protections against the login-over-email page. My own sanity checking also confirms everything still in working order - should be good to go now!

tgxworld commented:
Thank you @featheredtoast for the work! Merged in https://github.com/discourse/discourse/pull/5612
